{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#check-overlap-mcpas\" data-toc-modified-id=\"check-overlap-mcpas-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>check overlap mcpas</a></span><ul class=\"toc-item\"><li><span><a href=\"#VDJdb-update\" data-toc-modified-id=\"VDJdb-update-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>VDJdb update</a></span></li><li><span><a href=\"#10x-data\" data-toc-modified-id=\"10x-data-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>10x data</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creation of external validation sets based on mcpas and the held-out 10x genomics data. Based on retrieving all unique epitopes (i.e. those not occurring in the VDJdb pairs used for training)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.lines as lines\n",
    "import seaborn as sns\n",
    "import scipy\n",
    "\n",
    "from src.scripts.evaluate.visualize import predict_variations\n",
    "from src.config import PROJECT_ROOT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pandas.io.json import json_normalize\n",
    "from pathlib import Path\n",
    "\n",
    "from src.config import PROJECT_ROOT\n",
    "from src.scripts.preprocessing import preprocess_vdjdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# check overlap mcpas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  import sys\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  \n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering reduced the number of unique pairs from 61047 to 19842.\n"
     ]
    }
   ],
   "source": [
    "# import dataset again to start fresh\n",
    "vdjdb_normal_path = PROJECT_ROOT / \"data/raw/vdjdb/vdjdb-2019-08-08/vdjdb.txt\"\n",
    "df = pd.read_csv(vdjdb_normal_path, sep=\"\\t\")\n",
    "\n",
    "# expand the json/dict-like columns\n",
    "df = (\n",
    "    df.join(json_normalize(df[\"method\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"method.\"))\n",
    "    .join(json_normalize(df[\"meta\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"meta.\"))\n",
    "    .join(json_normalize(df[\"cdr3fix\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"cdr3fix.\"))\n",
    "    .drop([\"method\", \"meta\", \"cdr3fix\"], axis=1)\n",
    ")\n",
    "\n",
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "\n",
    "# create column based on cdr3 and epitope sequences\n",
    "df[\"merged\"] = df[\"cdr3\"] + \"-\" + df[\"antigen.epitope\"]\n",
    "\n",
    "# create length columns for figures\n",
    "df[\"cdr3.len\"] = df[\"cdr3\"].str.len()\n",
    "df[\"antigen.epitope.len\"] = df[\"antigen.epitope\"].str.len()\n",
    "\n",
    "df_human_no10x = df.loc[(df[\"species\"] == \"HomoSapiens\") & \n",
    "#                                  (df[\"gene\"] == \"TRB\") & \n",
    "                                 (df[\"cdr3fix.good\"]) & \n",
    "                                 (df[\"mhc.class\"] == \"MHCI\") &\n",
    "                                 ~(df[\"reference.id\"].str.contains(\"10x\"))]\n",
    "\n",
    "df_human_no10x = df_human_no10x.loc[ (df_human_no10x[\"antigen.epitope\"].str.len() >= 8) &\n",
    "                                                 (df_human_no10x[\"antigen.epitope\"].str.len() <= 11) &\n",
    "                                                 (df_human_no10x[\"cdr3\"].str.len() >= 10) &\n",
    "                                                 (df_human_no10x[\"cdr3\"].str.len() <= 20)]\n",
    "\n",
    "df_human_no10x = df_human_no10x.drop_duplicates(columns)\n",
    "\n",
    "print(f\"Filtering reduced the number of unique pairs from {df.drop_duplicates(columns).shape[0]} to {df_human_no10x.shape[0]}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_human_no10x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas = pd.read_csv(\"/media/pieter/DATA/Wetenschap/Doctoraat/projects/deepTCR/deepTCR/data/raw/mcpas.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas[\"merged\"] = df_mcpas[\"CDR3_beta\"] + \"-\" + df_mcpas[\"HLA_peptide\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas = df_mcpas.loc[ (df_mcpas[\"HLA_peptide\"].str.len() >= 8) &\n",
    "                         (df_mcpas[\"HLA_peptide\"].str.len() <= 11) &\n",
    "                         (df_mcpas[\"CDR3_beta\"].str.len() >= 10) &\n",
    "                         (df_mcpas[\"CDR3_beta\"].str.len() <= 20)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas = df_mcpas.drop_duplicates([\"CDR3_beta\",\"HLA_peptide\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>cdr3.len</th>\n",
       "      <th>antigen.epitope.len</th>\n",
       "      <th>TRBV_gene</th>\n",
       "      <th>CDR3_beta</th>\n",
       "      <th>TRBJ_gene</th>\n",
       "      <th>HLA_peptide</th>\n",
       "      <th>Epitope species</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Source</th>\n",
       "      <th>merged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVTDDKIIF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>TRBV27</td>\n",
       "      <td>CASSLGSSYEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>YLEPGPVTA</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8752841</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSLGSSYEQYF-YLEPGPVTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVDSGGYQKVTF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ13*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>TRBV4</td>\n",
       "      <td>CASLAGQGYNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>SAYGEPRKL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASLAGQGYNEQF-SAYGEPRKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAGGDDKIIF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>TRBV12</td>\n",
       "      <td>CASLGAQNNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>AARAVFLAL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASLGAQNNEQF-AARAVFLAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVKDARLMF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ31*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>TRBV6</td>\n",
       "      <td>CASRLWFWALEAF</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>SAYGEPRKL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASRLWFWALEAF-SAYGEPRKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVGSDKIIF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ30*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>TRBV6</td>\n",
       "      <td>CASSYSTGDEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>AARAVFLAL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSYSTGDEQYF-AARAVFLAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1560</th>\n",
       "      <td>355</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYLPGQGDHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEMGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "      <td>TRBV27</td>\n",
       "      <td>CASSSGQEAF</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>TPRVTGGGAM</td>\n",
       "      <td>Cytomegalovirus(CMV)</td>\n",
       "      <td>PubMed ID: 25339770</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSGQEAF-TPRVTGGGAM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1561</th>\n",
       "      <td>0</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEAGQGFFSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEMGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>TRBV5-07</td>\n",
       "      <td>CASSSGTSGYEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>QIKVRVKMV</td>\n",
       "      <td>Cytomegalovirus(CMV)</td>\n",
       "      <td>PubMed ID: 25339770</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSGTSGYEQYF-QIKVRVKMV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1562</th>\n",
       "      <td>356</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVPSGAGSYQLTF</td>\n",
       "      <td>TRAV20*01</td>\n",
       "      <td>TRAJ28*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEMGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>TRBV27</td>\n",
       "      <td>CASSSGTSGYYNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>GLCTLVAML</td>\n",
       "      <td>EpsteinBarrvirus(EBV)</td>\n",
       "      <td>PubMed ID: 25339770</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSGTSGYYNEQF-GLCTLVAML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>356</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFEPGQGFYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEMGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>TRBV27</td>\n",
       "      <td>CASSSPRESTDTQYF</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>YSEHPTFTSQY</td>\n",
       "      <td>Cytomegalovirus(CMV)</td>\n",
       "      <td>PubMed ID: 25339770</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSPRESTDTQYF-YSEHPTFTSQY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1565</th>\n",
       "      <td>357</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSYEPGQVSHYSNQPQHF</td>\n",
       "      <td>TRBV13*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>FLKEMGGL</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>8</td>\n",
       "      <td>TRBV7-02</td>\n",
       "      <td>CASSSQRESSDTQYF</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>YSEHPTFTSQY</td>\n",
       "      <td>Cytomegalovirus(CMV)</td>\n",
       "      <td>PubMed ID: 25339770</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSQRESSDTQYF-YSEHPTFTSQY</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>916 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      complex.id gene                  cdr3       v.segm      j.segm  \\\n",
       "0              0  TRA            CAVTDDKIIF  TRAV12-2*01   TRAJ30*01   \n",
       "1              0  TRA         CAVDSGGYQKVTF  TRAV12-2*01   TRAJ13*01   \n",
       "2              0  TRA            CAGGDDKIIF  TRAV12-2*01   TRAJ30*01   \n",
       "3              0  TRA            CAVKDARLMF  TRAV12-2*01   TRAJ31*01   \n",
       "4              0  TRA            CAVGSDKIIF  TRAV12-2*01   TRAJ30*01   \n",
       "...          ...  ...                   ...          ...         ...   \n",
       "1560         355  TRB  CASSYLPGQGDHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "1561           0  TRB   CASSFEAGQGFFSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "1562         356  TRA        CAVPSGAGSYQLTF    TRAV20*01   TRAJ28*01   \n",
       "1563         356  TRB   CASSFEPGQGFYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "1565         357  TRB  CASSYEPGQVSHYSNQPQHF    TRBV13*01  TRBJ1-5*01   \n",
       "\n",
       "          species     mhc.a mhc.b mhc.class antigen.epitope  ... cdr3.len  \\\n",
       "0     HomoSapiens  HLA-A*02   B2M      MHCI       LLWNGPMAV  ...       10   \n",
       "1     HomoSapiens  HLA-A*02   B2M      MHCI       LLWNGPMAV  ...       13   \n",
       "2     HomoSapiens  HLA-A*02   B2M      MHCI       LLWNGPMAV  ...       10   \n",
       "3     HomoSapiens  HLA-A*02   B2M      MHCI       LLWNGPMAV  ...       10   \n",
       "4     HomoSapiens  HLA-A*02   B2M      MHCI       LLWNGPMAV  ...       10   \n",
       "...           ...       ...   ...       ...             ...  ...      ...   \n",
       "1560  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEMGGL  ...       20   \n",
       "1561  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEMGGL  ...       19   \n",
       "1562  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEMGGL  ...       14   \n",
       "1563  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEMGGL  ...       19   \n",
       "1565  HomoSapiens  HLA-B*08   B2M      MHCI        FLKEMGGL  ...       20   \n",
       "\n",
       "     antigen.epitope.len TRBV_gene        CDR3_beta TRBJ_gene  HLA_peptide  \\\n",
       "0                      9    TRBV27    CASSLGSSYEQYF   TRBJ2-7    YLEPGPVTA   \n",
       "1                      9     TRBV4    CASLAGQGYNEQF   TRBJ2-1    SAYGEPRKL   \n",
       "2                      9    TRBV12     CASLGAQNNEQF   TRBJ2-1    AARAVFLAL   \n",
       "3                      9     TRBV6    CASRLWFWALEAF   TRBJ1-1    SAYGEPRKL   \n",
       "4                      9     TRBV6    CASSYSTGDEQYF   TRBJ2-7    AARAVFLAL   \n",
       "...                  ...       ...              ...       ...          ...   \n",
       "1560                   8    TRBV27       CASSSGQEAF   TRBJ1-1   TPRVTGGGAM   \n",
       "1561                   8  TRBV5-07   CASSSGTSGYEQYF   TRBJ2-7    QIKVRVKMV   \n",
       "1562                   8    TRBV27  CASSSGTSGYYNEQF   TRBJ2-1    GLCTLVAML   \n",
       "1563                   8    TRBV27  CASSSPRESTDTQYF   TRBJ2-3  YSEHPTFTSQY   \n",
       "1565                   8  TRBV7-02  CASSSQRESSDTQYF   TRBJ2-3  YSEHPTFTSQY   \n",
       "\n",
       "            Epitope species            Reference     Source  \\\n",
       "0                  Melanoma   PubMed ID: 8752841  McPas-TCR   \n",
       "1                  Melanoma   PubMed ID: 8921424  McPas-TCR   \n",
       "2                  Melanoma   PubMed ID: 8921424  McPas-TCR   \n",
       "3                  Melanoma   PubMed ID: 8921424  McPas-TCR   \n",
       "4                  Melanoma   PubMed ID: 8921424  McPas-TCR   \n",
       "...                     ...                  ...        ...   \n",
       "1560   Cytomegalovirus(CMV)  PubMed ID: 25339770  McPas-TCR   \n",
       "1561   Cytomegalovirus(CMV)  PubMed ID: 25339770  McPas-TCR   \n",
       "1562  EpsteinBarrvirus(EBV)  PubMed ID: 25339770  McPas-TCR   \n",
       "1563   Cytomegalovirus(CMV)  PubMed ID: 25339770  McPas-TCR   \n",
       "1565   Cytomegalovirus(CMV)  PubMed ID: 25339770  McPas-TCR   \n",
       "\n",
       "                           merged  \n",
       "0         CASSLGSSYEQYF-YLEPGPVTA  \n",
       "1         CASLAGQGYNEQF-SAYGEPRKL  \n",
       "2          CASLGAQNNEQF-AARAVFLAL  \n",
       "3         CASRLWFWALEAF-SAYGEPRKL  \n",
       "4         CASSYSTGDEQYF-AARAVFLAL  \n",
       "...                           ...  \n",
       "1560        CASSSGQEAF-TPRVTGGGAM  \n",
       "1561     CASSSGTSGYEQYF-QIKVRVKMV  \n",
       "1562    CASSSGTSGYYNEQF-GLCTLVAML  \n",
       "1563  CASSSPRESTDTQYF-YSEHPTFTSQY  \n",
       "1565  CASSSQRESSDTQYF-YSEHPTFTSQY  \n",
       "\n",
       "[916 rows x 65 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged = pd.concat([df,df_mcpas], axis=1, join='inner')\n",
    "merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas = df_mcpas.rename(columns={\"CDR3_beta\": \"cdr3\", \"HLA_peptide\": \"antigen.epitope\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cdr3</th>\n",
       "      <th>antigen.epitope</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CASSAGTGAYEQYF</td>\n",
       "      <td>LLWNGPMAV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CSARDRTGNGYTF</td>\n",
       "      <td>GLCTLVAML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CSVGTGGTNEKLFF</td>\n",
       "      <td>GLCTLVAML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CSVGSGGTNEKLFF</td>\n",
       "      <td>GLCTLVAML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CSVGAGGTNEKLFF</td>\n",
       "      <td>GLCTLVAML</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2511</th>\n",
       "      <td>CASSIISVDGYTF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2512</th>\n",
       "      <td>CASSTRSGTEQYF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513</th>\n",
       "      <td>CASSATGSTYEQYF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2514</th>\n",
       "      <td>CASSARATDTQYF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2515</th>\n",
       "      <td>CASSMFALDTQYF</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2516 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                cdr3 antigen.epitope\n",
       "0     CASSAGTGAYEQYF       LLWNGPMAV\n",
       "1      CSARDRTGNGYTF       GLCTLVAML\n",
       "2     CSVGTGGTNEKLFF       GLCTLVAML\n",
       "3     CSVGSGGTNEKLFF       GLCTLVAML\n",
       "4     CSVGAGGTNEKLFF       GLCTLVAML\n",
       "...              ...             ...\n",
       "2511   CASSIISVDGYTF       GILGFVFTL\n",
       "2512   CASSTRSGTEQYF       GILGFVFTL\n",
       "2513  CASSATGSTYEQYF       GILGFVFTL\n",
       "2514   CASSARATDTQYF       GILGFVFTL\n",
       "2515   CASSMFALDTQYF       GILGFVFTL\n",
       "\n",
       "[2516 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intersection = df[[\"cdr3\",\"antigen.epitope\"]].merge(df_mcpas[[\"cdr3\",\"antigen.epitope\"]]).drop_duplicates()\n",
    "intersection\n",
    "# pd.concat([df.merge(intersection), df_mcpas.merge(intersection)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_vdjdb = df_mcpas[\"merged\"].apply(lambda x: df[\"merged\"].str.contains(x).any())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0       False\n",
       "1       False\n",
       "2       False\n",
       "3       False\n",
       "4       False\n",
       "        ...  \n",
       "6680    False\n",
       "6681    False\n",
       "6682     True\n",
       "6684    False\n",
       "6685    False\n",
       "Name: merged, Length: 6324, dtype: bool"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vdjdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2516"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_vdjdb.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRBV_gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>TRBJ_gene</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>Epitope species</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Source</th>\n",
       "      <th>merged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRBV27</td>\n",
       "      <td>CASSLGSSYEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>YLEPGPVTA</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8752841</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSLGSSYEQYF-YLEPGPVTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TRBV4</td>\n",
       "      <td>CASLAGQGYNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>SAYGEPRKL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASLAGQGYNEQF-SAYGEPRKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRBV12</td>\n",
       "      <td>CASLGAQNNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>AARAVFLAL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASLGAQNNEQF-AARAVFLAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRBV6</td>\n",
       "      <td>CASRLWFWALEAF</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>SAYGEPRKL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASRLWFWALEAF-SAYGEPRKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRBV6</td>\n",
       "      <td>CASSYSTGDEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>AARAVFLAL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSYSTGDEQYF-AARAVFLAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6677</th>\n",
       "      <td>TRBV19</td>\n",
       "      <td>CASSKRSNPPQHF</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSKRSNPPQHF-GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6680</th>\n",
       "      <td>TRBV19</td>\n",
       "      <td>GASSIGIFGYTF</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>GASSIGIFGYTF-GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6681</th>\n",
       "      <td>TRBV19</td>\n",
       "      <td>CASRIGIYGYPF</td>\n",
       "      <td>TRBJ1-2</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASRIGIYGYPF-GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6684</th>\n",
       "      <td>TRBV19</td>\n",
       "      <td>CASSSRSAIEQFF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSRSAIEQFF-GILGFVFTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6685</th>\n",
       "      <td>TRBV7-2</td>\n",
       "      <td>CASSTSLGGGGRQSF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>GILGFVFTL</td>\n",
       "      <td>Influenza</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSTSLGGGGRQSF-GILGFVFTL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3808 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TRBV_gene             cdr3 TRBJ_gene antigen.epitope Epitope species  \\\n",
       "0       TRBV27    CASSLGSSYEQYF   TRBJ2-7       YLEPGPVTA        Melanoma   \n",
       "1        TRBV4    CASLAGQGYNEQF   TRBJ2-1       SAYGEPRKL        Melanoma   \n",
       "2       TRBV12     CASLGAQNNEQF   TRBJ2-1       AARAVFLAL        Melanoma   \n",
       "3        TRBV6    CASRLWFWALEAF   TRBJ1-1       SAYGEPRKL        Melanoma   \n",
       "4        TRBV6    CASSYSTGDEQYF   TRBJ2-7       AARAVFLAL        Melanoma   \n",
       "...        ...              ...       ...             ...             ...   \n",
       "6677    TRBV19    CASSKRSNPPQHF   TRBJ1-5       GILGFVFTL       Influenza   \n",
       "6680    TRBV19     GASSIGIFGYTF   TRBJ1-2       GILGFVFTL       Influenza   \n",
       "6681    TRBV19     CASRIGIYGYPF   TRBJ1-2       GILGFVFTL       Influenza   \n",
       "6684    TRBV19    CASSSRSAIEQFF   TRBJ2-1       GILGFVFTL       Influenza   \n",
       "6685   TRBV7-2  CASSTSLGGGGRQSF   TRBJ2-7       GILGFVFTL       Influenza   \n",
       "\n",
       "                Reference     Source                     merged  \n",
       "0      PubMed ID: 8752841  McPas-TCR    CASSLGSSYEQYF-YLEPGPVTA  \n",
       "1      PubMed ID: 8921424  McPas-TCR    CASLAGQGYNEQF-SAYGEPRKL  \n",
       "2      PubMed ID: 8921424  McPas-TCR     CASLGAQNNEQF-AARAVFLAL  \n",
       "3      PubMed ID: 8921424  McPas-TCR    CASRLWFWALEAF-SAYGEPRKL  \n",
       "4      PubMed ID: 8921424  McPas-TCR    CASSYSTGDEQYF-AARAVFLAL  \n",
       "...                   ...        ...                        ...  \n",
       "6677  PubMed ID: 28300170  McPas-TCR    CASSKRSNPPQHF-GILGFVFTL  \n",
       "6680  PubMed ID: 28300170  McPas-TCR     GASSIGIFGYTF-GILGFVFTL  \n",
       "6681  PubMed ID: 28300170  McPas-TCR     CASRIGIYGYPF-GILGFVFTL  \n",
       "6684  PubMed ID: 28300170  McPas-TCR    CASSSRSAIEQFF-GILGFVFTL  \n",
       "6685  PubMed ID: 28300170  McPas-TCR  CASSTSLGGGGRQSF-GILGFVFTL  \n",
       "\n",
       "[3808 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mcpas[~in_vdjdb]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['YLEPGPVTA', 'SAYGEPRKL', 'AARAVFLAL', 'EAAGIGILTV', 'IVTDFSVIK',\n",
       "       'AVFDRKSDAK', 'GLCTLVAML', 'RAKFKQLL', 'NLVPMVATV', 'EENLLDFVRF',\n",
       "       'KAFSPEVIPMF', 'RAFSPEVIPMF', 'RPRGEVRFL', 'ELAGIGILTV',\n",
       "       'HSKKKCDEL', 'ATDALMTGY', 'TPRVTGGGAM', 'FLRGRAYGL', 'TQGYFPDWQNY',\n",
       "       'EIYKRWII', 'FLKEKGGL', 'KRWIIMGLNK', 'ISPRTLNAW', 'QASQEVKNW',\n",
       "       'GLNKIVRMY', 'KRWIILGLNK', 'QYDPVAALF', 'YVLDHLIVV', 'FWIDLFETIG',\n",
       "       'RYPLTFGWCF', 'RYPLTFGW', 'GILGFVFTL', 'YSEHPTFTSQY', 'QIKVRVKMV',\n",
       "       'HPVGEADYFEY', 'RPPIFIRRL', 'EPLPQGQLTAY', 'VLEETSVML',\n",
       "       'ALTPVVVTL', 'FPRPWLHGL', 'IIKDYGKQM', 'LPPIVAKEI', 'HPKVSSEVHI',\n",
       "       'TPQDLNTML', 'TPGPGVRYPL', 'PQPELPYPQPE', 'LLWNGPMAV', 'VTEHDTLLY',\n",
       "       'LPRRSGAAGA', 'VLFGLGFAI'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mcpas.loc[~in_vdjdb,\"antigen.epitope\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_mcpas.loc[~in_vdjdb,\"antigen.epitope\"].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['YLEPGPVTA', 'SAYGEPRKL', 'AARAVFLAL', 'EAAGIGILTV', 'IVTDFSVIK',\n",
       "       'AVFDRKSDAK', 'GLCTLVAML', 'RAKFKQLL', 'NLVPMVATV', 'EENLLDFVRF',\n",
       "       'KAFSPEVIPMF', 'RAFSPEVIPMF', 'RPRGEVRFL', 'ELAGIGILTV',\n",
       "       'HSKKKCDEL', 'ATDALMTGY', 'TPRVTGGGAM', 'FLRGRAYGL', 'TQGYFPDWQNY',\n",
       "       'EIYKRWII', 'FLKEKGGL', 'KRWIIMGLNK', 'ISPRTLNAW', 'QASQEVKNW',\n",
       "       'GLNKIVRMY', 'KRWIILGLNK', 'QYDPVAALF', 'YVLDHLIVV', 'FWIDLFETIG',\n",
       "       'RYPLTFGWCF', 'RYPLTFGW', 'GILGFVFTL', 'YSEHPTFTSQY', 'QIKVRVKMV',\n",
       "       'HPVGEADYFEY', 'RPPIFIRRL', 'EPLPQGQLTAY', 'VLEETSVML',\n",
       "       'ALTPVVVTL', 'FPRPWLHGL', 'IIKDYGKQM', 'LPPIVAKEI', 'HPKVSSEVHI',\n",
       "       'TPQDLNTML', 'TPGPGVRYPL', 'PQPELPYPQPE', 'LLWNGPMAV', 'VTEHDTLLY',\n",
       "       'LPRRSGAAGA', 'VLFGLGFAI'], dtype=object)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mcpas.loc[~in_vdjdb,\"antigen.epitope\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AARAVFLAL',\n",
       " 'ALTPVVVTL',\n",
       " 'FWIDLFETIG',\n",
       " 'PQPELPYPQPE',\n",
       " 'RAFSPEVIPMF',\n",
       " 'RYPLTFGW',\n",
       " 'SAYGEPRKL',\n",
       " 'VLFGLGFAI',\n",
       " 'YLEPGPVTA'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_epitopes = set(df_mcpas.loc[~in_vdjdb,\"antigen.epitope\"].unique()) - set(df[\"antigen.epitope\"].unique())\n",
    "unique_epitopes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(64,)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mcpas.loc[df_mcpas[\"antigen.epitope\"].isin(unique_epitopes),\"antigen.epitope\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AARAVFLAL',\n",
       " 'ALTPVVVTL',\n",
       " 'FWIDLFETIG',\n",
       " 'PQPELPYPQPE',\n",
       " 'RAFSPEVIPMF',\n",
       " 'RYPLTFGW',\n",
       " 'SAYGEPRKL',\n",
       " 'VLFGLGFAI',\n",
       " 'YLEPGPVTA'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique = set(df_mcpas[\"antigen.epitope\"].unique()) - set(df[\"antigen.epitope\"].unique())\n",
    "unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRBV_gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>TRBJ_gene</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>Epitope species</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Source</th>\n",
       "      <th>merged</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TRBV27</td>\n",
       "      <td>CASSLGSSYEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>YLEPGPVTA</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8752841</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSLGSSYEQYF-YLEPGPVTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TRBV4</td>\n",
       "      <td>CASLAGQGYNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>SAYGEPRKL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASLAGQGYNEQF-SAYGEPRKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TRBV12</td>\n",
       "      <td>CASLGAQNNEQF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>AARAVFLAL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASLGAQNNEQF-AARAVFLAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TRBV6</td>\n",
       "      <td>CASRLWFWALEAF</td>\n",
       "      <td>TRBJ1-1</td>\n",
       "      <td>SAYGEPRKL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASRLWFWALEAF-SAYGEPRKL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TRBV6</td>\n",
       "      <td>CASSYSTGDEQYF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>AARAVFLAL</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>PubMed ID: 8921424</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSYSTGDEQYF-AARAVFLAL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6528</th>\n",
       "      <td>TRBV28</td>\n",
       "      <td>CASSSFVIAGGDYEQNF</td>\n",
       "      <td>TRBJ2-7</td>\n",
       "      <td>VLFGLGFAI</td>\n",
       "      <td>DiabetesType1</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSSFVIAGGDYEQNF-VLFGLGFAI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6529</th>\n",
       "      <td>TRBV3-1</td>\n",
       "      <td>CASRQDRWEKMGQNIRRF</td>\n",
       "      <td>TRBJ2-4</td>\n",
       "      <td>VLFGLGFAI</td>\n",
       "      <td>DiabetesType1</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASRQDRWEKMGQNIRRF-VLFGLGFAI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6530</th>\n",
       "      <td>TRBV6-5</td>\n",
       "      <td>CASSPVPGLDEQFF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>VLFGLGFAI</td>\n",
       "      <td>DiabetesType1</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CASSPVPGLDEQFF-VLFGLGFAI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6531</th>\n",
       "      <td>TRBV24-1</td>\n",
       "      <td>CATSDLPYIGTRLNEQFF</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>VLFGLGFAI</td>\n",
       "      <td>DiabetesType1</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CATSDLPYIGTRLNEQFF-VLFGLGFAI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6532</th>\n",
       "      <td>TRBV25</td>\n",
       "      <td>CAGLGDSGGSNYKLTF</td>\n",
       "      <td>TRBJ53</td>\n",
       "      <td>VLFGLGFAI</td>\n",
       "      <td>DiabetesType1</td>\n",
       "      <td>PubMed ID: 28300170</td>\n",
       "      <td>McPas-TCR</td>\n",
       "      <td>CAGLGDSGGSNYKLTF-VLFGLGFAI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>64 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     TRBV_gene                cdr3 TRBJ_gene antigen.epitope Epitope species  \\\n",
       "0       TRBV27       CASSLGSSYEQYF   TRBJ2-7       YLEPGPVTA        Melanoma   \n",
       "1        TRBV4       CASLAGQGYNEQF   TRBJ2-1       SAYGEPRKL        Melanoma   \n",
       "2       TRBV12        CASLGAQNNEQF   TRBJ2-1       AARAVFLAL        Melanoma   \n",
       "3        TRBV6       CASRLWFWALEAF   TRBJ1-1       SAYGEPRKL        Melanoma   \n",
       "4        TRBV6       CASSYSTGDEQYF   TRBJ2-7       AARAVFLAL        Melanoma   \n",
       "...        ...                 ...       ...             ...             ...   \n",
       "6528    TRBV28   CASSSFVIAGGDYEQNF   TRBJ2-7       VLFGLGFAI   DiabetesType1   \n",
       "6529   TRBV3-1  CASRQDRWEKMGQNIRRF   TRBJ2-4       VLFGLGFAI   DiabetesType1   \n",
       "6530   TRBV6-5      CASSPVPGLDEQFF   TRBJ2-1       VLFGLGFAI   DiabetesType1   \n",
       "6531  TRBV24-1  CATSDLPYIGTRLNEQFF   TRBJ2-1       VLFGLGFAI   DiabetesType1   \n",
       "6532    TRBV25    CAGLGDSGGSNYKLTF    TRBJ53       VLFGLGFAI   DiabetesType1   \n",
       "\n",
       "                Reference     Source                        merged  \n",
       "0      PubMed ID: 8752841  McPas-TCR       CASSLGSSYEQYF-YLEPGPVTA  \n",
       "1      PubMed ID: 8921424  McPas-TCR       CASLAGQGYNEQF-SAYGEPRKL  \n",
       "2      PubMed ID: 8921424  McPas-TCR        CASLGAQNNEQF-AARAVFLAL  \n",
       "3      PubMed ID: 8921424  McPas-TCR       CASRLWFWALEAF-SAYGEPRKL  \n",
       "4      PubMed ID: 8921424  McPas-TCR       CASSYSTGDEQYF-AARAVFLAL  \n",
       "...                   ...        ...                           ...  \n",
       "6528  PubMed ID: 28300170  McPas-TCR   CASSSFVIAGGDYEQNF-VLFGLGFAI  \n",
       "6529  PubMed ID: 28300170  McPas-TCR  CASRQDRWEKMGQNIRRF-VLFGLGFAI  \n",
       "6530  PubMed ID: 28300170  McPas-TCR      CASSPVPGLDEQFF-VLFGLGFAI  \n",
       "6531  PubMed ID: 28300170  McPas-TCR  CATSDLPYIGTRLNEQFF-VLFGLGFAI  \n",
       "6532  PubMed ID: 28300170  McPas-TCR    CAGLGDSGGSNYKLTF-VLFGLGFAI  \n",
       "\n",
       "[64 rows x 8 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mcpas[df_mcpas[\"antigen.epitope\"].isin(unique)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas_unique = df_mcpas[df_mcpas[\"antigen.epitope\"].isin(unique)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mcpas_unique.to_csv(PROJECT_ROOT / \"data/interim/vdjdb-2019-08-08/mcpas-human-trb-mhci-size.csv\", index=False, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VDJdb update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  import sys\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  \n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering reduced the number of unique pairs from 61555 to 20218.\n"
     ]
    }
   ],
   "source": [
    "# import dataset again to start fresh\n",
    "vdjdb_normal_path = PROJECT_ROOT / \"data/raw/vdjdb/vdjdb-2020-01-20/vdjdb.txt\"\n",
    "df_2020 = pd.read_csv(vdjdb_normal_path, sep=\"\\t\")\n",
    "\n",
    "# expand the json/dict-like columns\n",
    "df_2020 = (\n",
    "    df_2020.join(json_normalize(df_2020[\"method\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"method.\"))\n",
    "    .join(json_normalize(df_2020[\"meta\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"meta.\"))\n",
    "    .join(json_normalize(df_2020[\"cdr3fix\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"cdr3fix.\"))\n",
    "    .drop([\"method\", \"meta\", \"cdr3fix\"], axis=1)\n",
    ")\n",
    "\n",
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "\n",
    "# create column based on cdr3 and epitope sequences\n",
    "df_2020[\"merged\"] = df_2020[\"cdr3\"] + \"-\" + df_2020[\"antigen.epitope\"]\n",
    "\n",
    "# create length columns for figures\n",
    "df_2020[\"cdr3.len\"] = df_2020[\"cdr3\"].str.len()\n",
    "df_2020[\"antigen.epitope.len\"] = df_2020[\"antigen.epitope\"].str.len()\n",
    "\n",
    "df_2020_human_no10x = df_2020.loc[(df_2020[\"species\"] == \"HomoSapiens\") & \n",
    "#                                  (df_2020[\"gene\"] == \"TRB\") & \n",
    "                                 (df_2020[\"cdr3fix.good\"]) & \n",
    "                                 (df_2020[\"mhc.class\"] == \"MHCI\") &\n",
    "                                 ~(df_2020[\"reference.id\"].str.contains(\"10x\"))]\n",
    "\n",
    "# df_2020_human_no10x = df_2020_human_no10x.loc[ (df_2020_human_no10x[\"antigen.epitope\"].str.len() >= 8) &\n",
    "#                                                  (df_2020_human_no10x[\"antigen.epitope\"].str.len() <= 11) &\n",
    "#                                                  (df_2020_human_no10x[\"cdr3\"].str.len() >= 10) &\n",
    "#                                                  (df_2020_human_no10x[\"cdr3\"].str.len() <= 20)]\n",
    "\n",
    "df_2020_human_no10x = df_2020_human_no10x.drop_duplicates(columns)\n",
    "\n",
    "print(f\"Filtering reduced the number of unique pairs from {df_2020.drop_duplicates(columns).shape[0]} to {df_2020_human_no10x.shape[0]}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_2020 = df_2020_human_no10x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>cdr3fix.vId</th>\n",
       "      <th>cdr3fix.oldVEnd</th>\n",
       "      <th>cdr3fix.oldVFixType</th>\n",
       "      <th>cdr3fix.oldVId</th>\n",
       "      <th>cdr3fix.oldJFixType</th>\n",
       "      <th>cdr3fix.oldJStart</th>\n",
       "      <th>cdr3fix.oldJId</th>\n",
       "      <th>merged</th>\n",
       "      <th>cdr3.len</th>\n",
       "      <th>antigen.epitope.len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53237</th>\n",
       "      <td>20790</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CALSGFYNTDKLIF</td>\n",
       "      <td>TRAV19*01</td>\n",
       "      <td>TRAJ34*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*35:08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LPEPLPQGQLTAY</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV19*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CALSGFYNTDKLIF-LPEPLPQGQLTAY</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53238</th>\n",
       "      <td>20790</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSIGTGGSQPQHF</td>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*35:08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LPEPLPQGQLTAY</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV7-2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSIGTGGSQPQHF-LPEPLPQGQLTAY</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53240</th>\n",
       "      <td>20791</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSKLGTSEETQYF</td>\n",
       "      <td>TRBV5-6*01</td>\n",
       "      <td>TRBJ2-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*35:08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LPEPLPQGQLTAY</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV5-6*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSKLGTSEETQYF-LPEPLPQGQLTAY</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53242</th>\n",
       "      <td>20792</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASPGLAGEYEQYF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*35:08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LPEPLPQGQLTAY</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASPGLAGEYEQYF-LPEPLPQGQLTAY</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53244</th>\n",
       "      <td>20793</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASPGETEAFF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-B*35:08</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>LPEPLPQGQLTAY</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASPGETEAFF-LPEPLPQGQLTAY</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76146</th>\n",
       "      <td>24634</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CLVGDLPSGGYQKVTF</td>\n",
       "      <td>TRAV4*01</td>\n",
       "      <td>TRAJ13*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*68:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>DATYQRTRALVR</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV4*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLVGDLPSGGYQKVTF-DATYQRTRALVR</td>\n",
       "      <td>16</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76147</th>\n",
       "      <td>24634</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CSARDRGREKLFF</td>\n",
       "      <td>TRBV20-1*01</td>\n",
       "      <td>TRBJ1-4*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*68:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>DATYQRTRALVR</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV20-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CSARDRGREKLFF-DATYQRTRALVR</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76148</th>\n",
       "      <td>24635</td>\n",
       "      <td>TRA</td>\n",
       "      <td>CAVNNNNDMRF</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>TRAJ43*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*68:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>DATYQRTRALVR</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV12-2*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CAVNNNNDMRF-DATYQRTRALVR</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76149</th>\n",
       "      <td>24635</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CSGSQDPYEQYF</td>\n",
       "      <td>TRBV20-1*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*68:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>DATYQRTRALVR</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV20-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CSGSQDPYEQYF-DATYQRTRALVR</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76151</th>\n",
       "      <td>24636</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CTAEAGNTEAFF</td>\n",
       "      <td>TRBV7-2*03</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*68:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>DATYQRTRALVR</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV7-2*03</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>FailedNoAlignment</td>\n",
       "      <td>TRBV20-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTAEAGNTEAFF-DATYQRTRALVR</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>227 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       complex.id gene              cdr3       v.segm      j.segm  \\\n",
       "53237       20790  TRA    CALSGFYNTDKLIF    TRAV19*01   TRAJ34*01   \n",
       "53238       20790  TRB   CASSIGTGGSQPQHF   TRBV7-2*01  TRBJ1-5*01   \n",
       "53240       20791  TRB   CASSKLGTSEETQYF   TRBV5-6*01  TRBJ2-5*01   \n",
       "53242       20792  TRB    CASPGLAGEYEQYF   TRBV6-1*01  TRBJ2-7*01   \n",
       "53244       20793  TRB       CASPGETEAFF   TRBV6-1*01  TRBJ1-1*01   \n",
       "...           ...  ...               ...          ...         ...   \n",
       "76146       24634  TRA  CLVGDLPSGGYQKVTF     TRAV4*01   TRAJ13*01   \n",
       "76147       24634  TRB     CSARDRGREKLFF  TRBV20-1*01  TRBJ1-4*01   \n",
       "76148       24635  TRA       CAVNNNNDMRF  TRAV12-2*01   TRAJ43*01   \n",
       "76149       24635  TRB      CSGSQDPYEQYF  TRBV20-1*01  TRBJ2-7*01   \n",
       "76151       24636  TRB      CTAEAGNTEAFF   TRBV7-2*03  TRBJ1-1*01   \n",
       "\n",
       "           species        mhc.a mhc.b mhc.class antigen.epitope  ...  \\\n",
       "53237  HomoSapiens  HLA-B*35:08   B2M      MHCI   LPEPLPQGQLTAY  ...   \n",
       "53238  HomoSapiens  HLA-B*35:08   B2M      MHCI   LPEPLPQGQLTAY  ...   \n",
       "53240  HomoSapiens  HLA-B*35:08   B2M      MHCI   LPEPLPQGQLTAY  ...   \n",
       "53242  HomoSapiens  HLA-B*35:08   B2M      MHCI   LPEPLPQGQLTAY  ...   \n",
       "53244  HomoSapiens  HLA-B*35:08   B2M      MHCI   LPEPLPQGQLTAY  ...   \n",
       "...            ...          ...   ...       ...             ...  ...   \n",
       "76146  HomoSapiens  HLA-A*68:01   B2M      MHCI    DATYQRTRALVR  ...   \n",
       "76147  HomoSapiens  HLA-A*68:01   B2M      MHCI    DATYQRTRALVR  ...   \n",
       "76148  HomoSapiens  HLA-A*68:01   B2M      MHCI    DATYQRTRALVR  ...   \n",
       "76149  HomoSapiens  HLA-A*68:01   B2M      MHCI    DATYQRTRALVR  ...   \n",
       "76151  HomoSapiens  HLA-A*68:01   B2M      MHCI    DATYQRTRALVR  ...   \n",
       "\n",
       "       cdr3fix.vId cdr3fix.oldVEnd cdr3fix.oldVFixType  cdr3fix.oldVId  \\\n",
       "53237    TRAV19*01             NaN                 NaN             NaN   \n",
       "53238   TRBV7-2*01             NaN                 NaN             NaN   \n",
       "53240   TRBV5-6*01             NaN                 NaN             NaN   \n",
       "53242   TRBV6-1*01             NaN                 NaN             NaN   \n",
       "53244   TRBV6-1*01             NaN                 NaN             NaN   \n",
       "...            ...             ...                 ...             ...   \n",
       "76146     TRAV4*01             NaN                 NaN             NaN   \n",
       "76147  TRBV20-1*01             NaN                 NaN             NaN   \n",
       "76148  TRAV12-2*01             NaN                 NaN             NaN   \n",
       "76149  TRBV20-1*01             NaN                 NaN             NaN   \n",
       "76151   TRBV7-2*03            -1.0   FailedNoAlignment     TRBV20-1*01   \n",
       "\n",
       "      cdr3fix.oldJFixType cdr3fix.oldJStart cdr3fix.oldJId  \\\n",
       "53237                 NaN               NaN            NaN   \n",
       "53238                 NaN               NaN            NaN   \n",
       "53240                 NaN               NaN            NaN   \n",
       "53242                 NaN               NaN            NaN   \n",
       "53244                 NaN               NaN            NaN   \n",
       "...                   ...               ...            ...   \n",
       "76146                 NaN               NaN            NaN   \n",
       "76147                 NaN               NaN            NaN   \n",
       "76148                 NaN               NaN            NaN   \n",
       "76149                 NaN               NaN            NaN   \n",
       "76151                 NaN               NaN            NaN   \n",
       "\n",
       "                              merged cdr3.len antigen.epitope.len  \n",
       "53237   CALSGFYNTDKLIF-LPEPLPQGQLTAY       14                  13  \n",
       "53238  CASSIGTGGSQPQHF-LPEPLPQGQLTAY       15                  13  \n",
       "53240  CASSKLGTSEETQYF-LPEPLPQGQLTAY       15                  13  \n",
       "53242   CASPGLAGEYEQYF-LPEPLPQGQLTAY       14                  13  \n",
       "53244      CASPGETEAFF-LPEPLPQGQLTAY       11                  13  \n",
       "...                              ...      ...                 ...  \n",
       "76146  CLVGDLPSGGYQKVTF-DATYQRTRALVR       16                  12  \n",
       "76147     CSARDRGREKLFF-DATYQRTRALVR       13                  12  \n",
       "76148       CAVNNNNDMRF-DATYQRTRALVR       11                  12  \n",
       "76149      CSGSQDPYEQYF-DATYQRTRALVR       12                  12  \n",
       "76151      CTAEAGNTEAFF-DATYQRTRALVR       12                  12  \n",
       "\n",
       "[227 rows x 57 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique = set(df_2020_human_no10x[\"antigen.epitope\"].unique()) - set(df[\"antigen.epitope\"].unique())\n",
    "df_2020_human_no10x[df_2020_human_no10x[\"antigen.epitope\"].isin(unique)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_2020_human_no10x.loc[df_2020_human_no10x[\"antigen.epitope\"].isin(unique), \"mhc.class\"].nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "New epitopes all fall outside of length restrictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "left_only     376\n",
       "both            0\n",
       "right_only      0\n",
       "Name: _merge, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "df_diff_unique = df_2020_human_no10x.drop_duplicates(columns).fillna(\"NULL\").merge(df[columns].drop_duplicates(columns).fillna(\"NULL\"),indicator = True, how='outer', on=columns).loc[lambda x : x['_merge']!='both']\n",
    "df_diff_unique[\"_merge\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "antigen.epitope\n",
       "DATYQRTRALVR       190\n",
       "NLVPMVATV           58\n",
       "GILGFVFTL           44\n",
       "LLLGIGILV           16\n",
       "ELAGIGILTV          12\n",
       "LPEPLPQGQLTAY        9\n",
       "CPSQEPMSIYVY         6\n",
       "LLWNGPMAV            3\n",
       "KRWIILGLNK           3\n",
       "ARNLVPMVATVQGQN      3\n",
       "LPEPLPQGGLTAY        2\n",
       "KLVALGINAV           2\n",
       "CINGVCWTV            2\n",
       "QVPLRPMTYK           2\n",
       "LSEFCRVLCCYVLEE      2\n",
       "LPEPLPQGQLTGY        2\n",
       "LPEPLPQAQLTAY        2\n",
       "GPEPLPQGQLTAY        2\n",
       "LPEPLPQGQLGAY        2\n",
       "LPEPLPQGQGTAY        2\n",
       "LPEGLPQGQLTAY        2\n",
       "LPEPLGQGQLTAY        2\n",
       "IVTDFSVIK            1\n",
       "IPSINVHHY            1\n",
       "GLCTLVAML            1\n",
       "SFHSLHLLF            1\n",
       "EDVPSGKLFMHVTLG      1\n",
       "NEGVKAAW             1\n",
       "CLGGLLTMV            1\n",
       "TPQDLNTML            1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diff_unique[df_diff_unique[\"cdr3fix.good\"]].drop_duplicates(columns).groupby(\"antigen.epitope\").size().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "antigen.epitope\n",
       "DATYQRTRALVR       190\n",
       "NLVPMVATV           58\n",
       "GILGFVFTL           44\n",
       "LLLGIGILV           16\n",
       "ELAGIGILTV          12\n",
       "LPEPLPQGQLTAY        9\n",
       "CPSQEPMSIYVY         6\n",
       "LLWNGPMAV            3\n",
       "KRWIILGLNK           3\n",
       "ARNLVPMVATVQGQN      3\n",
       "LPEPLPQGGLTAY        2\n",
       "KLVALGINAV           2\n",
       "CINGVCWTV            2\n",
       "QVPLRPMTYK           2\n",
       "LSEFCRVLCCYVLEE      2\n",
       "LPEPLPQGQLTGY        2\n",
       "LPEPLPQAQLTAY        2\n",
       "GPEPLPQGQLTAY        2\n",
       "LPEPLPQGQLGAY        2\n",
       "LPEPLPQGQGTAY        2\n",
       "LPEGLPQGQLTAY        2\n",
       "LPEPLGQGQLTAY        2\n",
       "IVTDFSVIK            1\n",
       "IPSINVHHY            1\n",
       "GLCTLVAML            1\n",
       "SFHSLHLLF            1\n",
       "EDVPSGKLFMHVTLG      1\n",
       "NEGVKAAW             1\n",
       "CLGGLLTMV            1\n",
       "TPQDLNTML            1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diff_unique[df_diff_unique[\"cdr3fix.good\"]].drop_duplicates(columns).groupby(\"antigen.epitope\").size().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  import sys\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  \n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  if __name__ == '__main__':\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:45: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:46: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:47: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n"
     ]
    }
   ],
   "source": [
    "# import dataset again to start fresh\n",
    "vdjdb_normal_path = PROJECT_ROOT / \"data/raw/vdjdb/vdjdb-2020-01-20/vdjdb.txt\"\n",
    "df = pd.read_csv(vdjdb_normal_path, sep=\"\\t\")\n",
    "\n",
    "# expand the json/dict-like columns\n",
    "df = (\n",
    "    df.join(json_normalize(df[\"method\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"method.\"))\n",
    "    .join(json_normalize(df[\"meta\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"meta.\"))\n",
    "    .join(json_normalize(df[\"cdr3fix\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"cdr3fix.\"))\n",
    "    .drop([\"method\", \"meta\", \"cdr3fix\"], axis=1)\n",
    ")\n",
    "\n",
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "\n",
    "# create column based on cdr3 and epitope sequences\n",
    "df[\"merged\"] = df[\"cdr3\"] + \"-\" + df[\"antigen.epitope\"]\n",
    "\n",
    "# create length columns for figures\n",
    "df[\"cdr3.len\"] = df[\"cdr3\"].str.len()\n",
    "df[\"antigen.epitope.len\"] = df[\"antigen.epitope\"].str.len()\n",
    "\n",
    "df = df.loc[(df[\"species\"] == \"HomoSapiens\") & \n",
    "#                                  (df_2020[\"gene\"] == \"TRB\") & \n",
    "                                 (df[\"cdr3fix.good\"]) & \n",
    "                                 (df[\"mhc.class\"] == \"MHCI\") &\n",
    "                                 ~(df[\"reference.id\"].str.contains(\"10x\"))]\n",
    "\n",
    "df = df.loc[ (df[\"antigen.epitope\"].str.len() >= 8) &\n",
    "                                                 (df[\"antigen.epitope\"].str.len() <= 11) &\n",
    "                                                 (df[\"cdr3\"].str.len() >= 10) &\n",
    "                                                 (df[\"cdr3\"].str.len() <= 20)]\n",
    "\n",
    "df = df.drop_duplicates(columns)\n",
    "\n",
    "df_jan = df\n",
    "\n",
    "###\n",
    "\n",
    "# import dataset again to start fresh\n",
    "vdjdb_normal_path = PROJECT_ROOT / \"data/raw/vdjdb/vdjdb-2019-08-08/vdjdb.txt\"\n",
    "df = pd.read_csv(vdjdb_normal_path, sep=\"\\t\")\n",
    "\n",
    "# expand the json/dict-like columns\n",
    "df = (\n",
    "    df.join(json_normalize(df[\"method\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"method.\"))\n",
    "    .join(json_normalize(df[\"meta\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"meta.\"))\n",
    "    .join(json_normalize(df[\"cdr3fix\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"cdr3fix.\"))\n",
    "    .drop([\"method\", \"meta\", \"cdr3fix\"], axis=1)\n",
    ")\n",
    "\n",
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "\n",
    "# create column based on cdr3 and epitope sequences\n",
    "df[\"merged\"] = df[\"cdr3\"] + \"-\" + df[\"antigen.epitope\"]\n",
    "\n",
    "# create length columns for figures\n",
    "df[\"cdr3.len\"] = df[\"cdr3\"].str.len()\n",
    "df[\"antigen.epitope.len\"] = df[\"antigen.epitope\"].str.len()\n",
    "\n",
    "\n",
    "df = df.loc[(df[\"species\"] == \"HomoSapiens\") & \n",
    "#                                  (df_2020[\"gene\"] == \"TRB\") & \n",
    "                                 (df[\"cdr3fix.good\"]) & \n",
    "                                 (df[\"mhc.class\"] == \"MHCI\") &\n",
    "                                 ~(df[\"reference.id\"].str.contains(\"10x\"))]\n",
    "\n",
    "df = df.loc[ (df[\"antigen.epitope\"].str.len() >= 8) &\n",
    "                                                 (df[\"antigen.epitope\"].str.len() <= 11) &\n",
    "                                                 (df[\"cdr3\"].str.len() >= 10) &\n",
    "                                                 (df[\"cdr3\"].str.len() <= 20)]\n",
    "\n",
    "df = df.drop_duplicates(columns)\n",
    "\n",
    "df_aug = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_diff = df_aug.fillna(\"NULL\").merge(df_jan.fillna(\"NULL\"),indicator = True, how='outer').loc[lambda x : x['_merge']!='both']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_diff_unique = df_jan.drop_duplicates(columns).fillna(\"NULL\").merge(df_aug[columns].drop_duplicates(columns).fillna(\"NULL\"),indicator = True, how='outer', on=columns).loc[lambda x : x['_merge']!='both']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "left_only     1\n",
       "both          0\n",
       "right_only    0\n",
       "Name: _merge, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diff_unique[\"_merge\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "antigen.epitope\n",
       "SFHSLHLLF    1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_diff_unique[df_diff_unique[\"cdr3fix.good\"]].drop_duplicates(columns).groupby(\"antigen.epitope\").size().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 10x data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  import sys\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  \n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering reduced the number of unique pairs from 61047 to 17200.\n"
     ]
    }
   ],
   "source": [
    "# import dataset again to start fresh\n",
    "vdjdb_normal_path = PROJECT_ROOT / \"data/raw/vdjdb/vdjdb-2019-08-08/vdjdb.txt\"\n",
    "df = pd.read_csv(vdjdb_normal_path, sep=\"\\t\")\n",
    "\n",
    "# expand the json/dict-like columns\n",
    "df = (\n",
    "    df.join(json_normalize(df[\"method\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"method.\"))\n",
    "    .join(json_normalize(df[\"meta\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"meta.\"))\n",
    "    .join(json_normalize(df[\"cdr3fix\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"cdr3fix.\"))\n",
    "    .drop([\"method\", \"meta\", \"cdr3fix\"], axis=1)\n",
    ")\n",
    "\n",
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "\n",
    "# create column based on cdr3 and epitope sequences\n",
    "df[\"merged\"] = df[\"cdr3\"] + \"-\" + df[\"antigen.epitope\"]\n",
    "\n",
    "# create length columns for figures\n",
    "df[\"cdr3.len\"] = df[\"cdr3\"].str.len()\n",
    "df[\"antigen.epitope.len\"] = df[\"antigen.epitope\"].str.len()\n",
    "\n",
    "df_human_10x = df.loc[(df[\"species\"] == \"HomoSapiens\") & \n",
    "                                 (df[\"gene\"] == \"TRB\") & \n",
    "                                 (df[\"cdr3fix.good\"]) & \n",
    "                                 (df[\"mhc.class\"] == \"MHCI\") &\n",
    "                                 (df[\"reference.id\"].str.contains(\"10x\"))]\n",
    "\n",
    "df_human_10x = df_human_10x.loc[ (df_human_10x[\"antigen.epitope\"].str.len() >= 8) &\n",
    "                                                 (df_human_10x[\"antigen.epitope\"].str.len() <= 11) &\n",
    "                                                 (df_human_10x[\"cdr3\"].str.len() >= 10) &\n",
    "                                                 (df_human_10x[\"cdr3\"].str.len() <= 20)]\n",
    "\n",
    "df_human_10x = df_human_10x.drop_duplicates(columns)\n",
    "\n",
    "print(f\"Filtering reduced the number of unique pairs from {df.drop_duplicates(columns).shape[0]} to {df_human_10x.shape[0]}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  import sys\n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:8: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  \n",
      "/home/pieter/miniconda3/envs/deepTCR/lib/python3.7/site-packages/ipykernel_launcher.py:9: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead\n",
      "  if __name__ == '__main__':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering reduced the number of unique pairs from 61047 to 19842.\n"
     ]
    }
   ],
   "source": [
    "# import dataset again to start fresh\n",
    "vdjdb_normal_path = PROJECT_ROOT / \"data/raw/vdjdb/vdjdb-2019-08-08/vdjdb.txt\"\n",
    "df = pd.read_csv(vdjdb_normal_path, sep=\"\\t\")\n",
    "\n",
    "# expand the json/dict-like columns\n",
    "df = (\n",
    "    df.join(json_normalize(df[\"method\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"method.\"))\n",
    "    .join(json_normalize(df[\"meta\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"meta.\"))\n",
    "    .join(json_normalize(df[\"cdr3fix\"].apply(lambda x: json.loads(r'{}'.format(x)))).add_prefix(\"cdr3fix.\"))\n",
    "    .drop([\"method\", \"meta\", \"cdr3fix\"], axis=1)\n",
    ")\n",
    "\n",
    "columns = [\"cdr3\", \"antigen.epitope\"]\n",
    "\n",
    "# create column based on cdr3 and epitope sequences\n",
    "df[\"merged\"] = df[\"cdr3\"] + \"-\" + df[\"antigen.epitope\"]\n",
    "\n",
    "# create length columns for figures\n",
    "df[\"cdr3.len\"] = df[\"cdr3\"].str.len()\n",
    "df[\"antigen.epitope.len\"] = df[\"antigen.epitope\"].str.len()\n",
    "\n",
    "df_human_no10x = df.loc[(df[\"species\"] == \"HomoSapiens\") & \n",
    "#                                  (df[\"gene\"] == \"TRB\") & \n",
    "                                 (df[\"cdr3fix.good\"]) & \n",
    "                                 (df[\"mhc.class\"] == \"MHCI\") &\n",
    "                                 ~(df[\"reference.id\"].str.contains(\"10x\"))]\n",
    "\n",
    "df_human_no10x = df_human_no10x.loc[ (df_human_no10x[\"antigen.epitope\"].str.len() >= 8) &\n",
    "                                                 (df_human_no10x[\"antigen.epitope\"].str.len() <= 11) &\n",
    "                                                 (df_human_no10x[\"cdr3\"].str.len() >= 10) &\n",
    "                                                 (df_human_no10x[\"cdr3\"].str.len() <= 20)]\n",
    "\n",
    "df_human_no10x = df_human_no10x.drop_duplicates(columns)\n",
    "\n",
    "print(f\"Filtering reduced the number of unique pairs from {df.drop_duplicates(columns).shape[0]} to {df_human_no10x.shape[0]}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>cdr3fix.vId</th>\n",
       "      <th>cdr3fix.oldVEnd</th>\n",
       "      <th>cdr3fix.oldVFixType</th>\n",
       "      <th>cdr3fix.oldVId</th>\n",
       "      <th>cdr3fix.oldJFixType</th>\n",
       "      <th>cdr3fix.oldJId</th>\n",
       "      <th>cdr3fix.oldJStart</th>\n",
       "      <th>merged</th>\n",
       "      <th>cdr3.len</th>\n",
       "      <th>antigen.epitope.len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1567</th>\n",
       "      <td>358</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSEGWHSYEQYF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSEGWHSYEQYF-KLGGALQAK</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569</th>\n",
       "      <td>359</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASGLNIDGDEQFF</td>\n",
       "      <td>TRBV12-5*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV12-5*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASGLNIDGDEQFF-KLGGALQAK</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1575</th>\n",
       "      <td>362</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSVMLDSPLHF</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>TRBJ1-6*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSVMLDSPLHF-KLGGALQAK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>363</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSLESGFLSGYTF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>CYTWNQMNL</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLESGFLSGYTF-CYTWNQMNL</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1579</th>\n",
       "      <td>364</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFTYRDTQYF</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFTYRDTQYF-KLGGALQAK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42267</th>\n",
       "      <td>20708</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSLGGGRPQHF</td>\n",
       "      <td>TRBV7-8*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV7-8*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLGGGRPQHF-KLGGALQAK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42271</th>\n",
       "      <td>20710</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSHTSADEQFF</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>RLRAEAQVK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSHTSADEQFF-RLRAEAQVK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42275</th>\n",
       "      <td>20712</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSWRTRDNQPQHF</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSWRTRDNQPQHF-KLGGALQAK</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42279</th>\n",
       "      <td>20714</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSQDVRGSNSPLHF</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>TRBJ1-6*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSQDVRGSNSPLHF-KLGGALQAK</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42281</th>\n",
       "      <td>20715</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASRVGPEAFF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASRVGPEAFF-KLGGALQAK</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13271 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       complex.id gene              cdr3       v.segm      j.segm  \\\n",
       "1567          358  TRB    CASSEGWHSYEQYF   TRBV6-1*01  TRBJ2-7*01   \n",
       "1569          359  TRB    CASGLNIDGDEQFF  TRBV12-5*01  TRBJ2-1*01   \n",
       "1575          362  TRB     CASSVMLDSPLHF     TRBV9*01  TRBJ1-6*01   \n",
       "1577          363  TRB   CASSLESGFLSGYTF   TRBV7-9*01  TRBJ1-2*01   \n",
       "1579          364  TRB     CASSFTYRDTQYF    TRBV28*01  TRBJ2-3*01   \n",
       "...           ...  ...               ...          ...         ...   \n",
       "42267       20708  TRB     CASSLGGGRPQHF   TRBV7-8*01  TRBJ1-5*01   \n",
       "42271       20710  TRB     CASSHTSADEQFF   TRBV4-3*01  TRBJ2-1*01   \n",
       "42275       20712  TRB   CASSWRTRDNQPQHF    TRBV28*01  TRBJ1-5*01   \n",
       "42279       20714  TRB  CASSQDVRGSNSPLHF   TRBV4-3*01  TRBJ1-6*01   \n",
       "42281       20715  TRB       CASRVGPEAFF   TRBV6-1*01  TRBJ1-1*01   \n",
       "\n",
       "           species        mhc.a mhc.b mhc.class antigen.epitope  ...  \\\n",
       "1567   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "1569   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "1575   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "1577   HomoSapiens  HLA-A*24:02   B2M      MHCI       CYTWNQMNL  ...   \n",
       "1579   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "...            ...          ...   ...       ...             ...  ...   \n",
       "42267  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "42271  HomoSapiens  HLA-A*03:01   B2M      MHCI       RLRAEAQVK  ...   \n",
       "42275  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "42279  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "42281  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "\n",
       "       cdr3fix.vId cdr3fix.oldVEnd cdr3fix.oldVFixType  cdr3fix.oldVId  \\\n",
       "1567    TRBV6-1*01             NaN                 NaN             NaN   \n",
       "1569   TRBV12-5*01             NaN                 NaN             NaN   \n",
       "1575      TRBV9*01             NaN                 NaN             NaN   \n",
       "1577    TRBV7-9*01             NaN                 NaN             NaN   \n",
       "1579     TRBV28*01             NaN                 NaN             NaN   \n",
       "...            ...             ...                 ...             ...   \n",
       "42267   TRBV7-8*01             NaN                 NaN             NaN   \n",
       "42271   TRBV4-3*01             NaN                 NaN             NaN   \n",
       "42275    TRBV28*01             NaN                 NaN             NaN   \n",
       "42279   TRBV4-3*01             NaN                 NaN             NaN   \n",
       "42281   TRBV6-1*01             NaN                 NaN             NaN   \n",
       "\n",
       "      cdr3fix.oldJFixType cdr3fix.oldJId cdr3fix.oldJStart  \\\n",
       "1567                  NaN            NaN               NaN   \n",
       "1569                  NaN            NaN               NaN   \n",
       "1575                  NaN            NaN               NaN   \n",
       "1577                  NaN            NaN               NaN   \n",
       "1579                  NaN            NaN               NaN   \n",
       "...                   ...            ...               ...   \n",
       "42267                 NaN            NaN               NaN   \n",
       "42271                 NaN            NaN               NaN   \n",
       "42275                 NaN            NaN               NaN   \n",
       "42279                 NaN            NaN               NaN   \n",
       "42281                 NaN            NaN               NaN   \n",
       "\n",
       "                           merged cdr3.len antigen.epitope.len  \n",
       "1567     CASSEGWHSYEQYF-KLGGALQAK       14                   9  \n",
       "1569     CASGLNIDGDEQFF-KLGGALQAK       14                   9  \n",
       "1575      CASSVMLDSPLHF-KLGGALQAK       13                   9  \n",
       "1577    CASSLESGFLSGYTF-CYTWNQMNL       15                   9  \n",
       "1579      CASSFTYRDTQYF-KLGGALQAK       13                   9  \n",
       "...                           ...      ...                 ...  \n",
       "42267     CASSLGGGRPQHF-KLGGALQAK       13                   9  \n",
       "42271     CASSHTSADEQFF-RLRAEAQVK       13                   9  \n",
       "42275   CASSWRTRDNQPQHF-KLGGALQAK       15                   9  \n",
       "42279  CASSQDVRGSNSPLHF-KLGGALQAK       16                   9  \n",
       "42281       CASRVGPEAFF-KLGGALQAK       11                   9  \n",
       "\n",
       "[13271 rows x 57 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_10x = set(df_human_10x[\"antigen.epitope\"].unique()) - set(df_human_no10x[\"antigen.epitope\"].unique())\n",
    "df_human_10x[df_human_10x[\"antigen.epitope\"].isin(unique_10x)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AYAQKIFKI',\n",
       " 'CLLGTYTQDV',\n",
       " 'CLLWSFQTSA',\n",
       " 'CYTWNQMNL',\n",
       " 'FLASKIGRLV',\n",
       " 'FLYALALLL',\n",
       " 'IMDQVPFSV',\n",
       " 'KLGGALQAK',\n",
       " 'KLQCVDLHV',\n",
       " 'KTWGQYWQV',\n",
       " 'KVAELVHFL',\n",
       " 'KVLEYVIKV',\n",
       " 'LLDFVRFMGV',\n",
       " 'MLDLQPETT',\n",
       " 'QPRAPIRPI',\n",
       " 'RIAAWMATY',\n",
       " 'RLRAEAQVK',\n",
       " 'RTLNAWVKV',\n",
       " 'SLFNTVATL',\n",
       " 'SLFNTVATLY',\n",
       " 'SLYNTVATLY',\n",
       " 'YLLEMLWRL',\n",
       " 'YLNDHLEPWI'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_10x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['KLGGALQAK', 'CYTWNQMNL', 'RIAAWMATY', 'SLFNTVATLY', 'RLRAEAQVK',\n",
       "       'LLDFVRFMGV', 'AYAQKIFKI', 'FLYALALLL', 'FLASKIGRLV', 'MLDLQPETT',\n",
       "       'IMDQVPFSV', 'YLLEMLWRL', 'KVLEYVIKV', 'RTLNAWVKV', 'KTWGQYWQV',\n",
       "       'YLNDHLEPWI', 'QPRAPIRPI', 'SLFNTVATL', 'CLLWSFQTSA', 'CLLGTYTQDV',\n",
       "       'KVAELVHFL', 'SLYNTVATLY', 'KLQCVDLHV'], dtype=object)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_human_10x.loc[df_human_10x[\"antigen.epitope\"].isin(unique_10x), \"antigen.epitope\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>cdr3fix.vId</th>\n",
       "      <th>cdr3fix.oldVEnd</th>\n",
       "      <th>cdr3fix.oldVFixType</th>\n",
       "      <th>cdr3fix.oldVId</th>\n",
       "      <th>cdr3fix.oldJFixType</th>\n",
       "      <th>cdr3fix.oldJId</th>\n",
       "      <th>cdr3fix.oldJStart</th>\n",
       "      <th>merged</th>\n",
       "      <th>cdr3.len</th>\n",
       "      <th>antigen.epitope.len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [complex.id, gene, cdr3, v.segm, j.segm, species, mhc.a, mhc.b, mhc.class, antigen.epitope, antigen.gene, antigen.species, reference.id, vdjdb.score, web.method, web.method.seq, web.cdr3fix.nc, web.cdr3fix.unmp, method.frequency, method.identification, method.sequencing, method.singlecell, method.verification, meta.cell.subset, meta.clone.id, meta.donor.MHC, meta.donor.MHC.method, meta.epitope.id, meta.replica.id, meta.samples.found, meta.structure.id, meta.studies.found, meta.study.id, meta.subject.cohort, meta.subject.id, meta.tissue, cdr3fix.cdr3, cdr3fix.cdr3_old, cdr3fix.fixNeeded, cdr3fix.good, cdr3fix.jCanonical, cdr3fix.jFixType, cdr3fix.jId, cdr3fix.jStart, cdr3fix.vCanonical, cdr3fix.vEnd, cdr3fix.vFixType, cdr3fix.vId, cdr3fix.oldVEnd, cdr3fix.oldVFixType, cdr3fix.oldVId, cdr3fix.oldJFixType, cdr3fix.oldJId, cdr3fix.oldJStart, merged, cdr3.len, antigen.epitope.len]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 57 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_human_no10x[df_human_no10x[\"antigen.epitope\"] == \"MLDLQPETT\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KLGGALQAK     12548\n",
       "RLRAEAQVK       407\n",
       "RTLNAWVKV        46\n",
       "AYAQKIFKI        39\n",
       "SLFNTVATLY       38\n",
       "FLASKIGRLV       34\n",
       "FLYALALLL        28\n",
       "LLDFVRFMGV       21\n",
       "KTWGQYWQV        16\n",
       "CYTWNQMNL        15\n",
       "MLDLQPETT        14\n",
       "YLLEMLWRL        13\n",
       "IMDQVPFSV        12\n",
       "KVLEYVIKV         8\n",
       "YLNDHLEPWI        7\n",
       "CLLWSFQTSA        7\n",
       "QPRAPIRPI         5\n",
       "SLFNTVATL         5\n",
       "CLLGTYTQDV        3\n",
       "SLYNTVATLY        2\n",
       "KVAELVHFL         1\n",
       "KLQCVDLHV         1\n",
       "RIAAWMATY         1\n",
       "Name: antigen.epitope, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_human_10x.loc[df_human_10x[\"antigen.epitope\"].isin(unique_10x), \"antigen.epitope\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complex.id</th>\n",
       "      <th>gene</th>\n",
       "      <th>cdr3</th>\n",
       "      <th>v.segm</th>\n",
       "      <th>j.segm</th>\n",
       "      <th>species</th>\n",
       "      <th>mhc.a</th>\n",
       "      <th>mhc.b</th>\n",
       "      <th>mhc.class</th>\n",
       "      <th>antigen.epitope</th>\n",
       "      <th>...</th>\n",
       "      <th>cdr3fix.vId</th>\n",
       "      <th>cdr3fix.oldVEnd</th>\n",
       "      <th>cdr3fix.oldVFixType</th>\n",
       "      <th>cdr3fix.oldVId</th>\n",
       "      <th>cdr3fix.oldJFixType</th>\n",
       "      <th>cdr3fix.oldJId</th>\n",
       "      <th>cdr3fix.oldJStart</th>\n",
       "      <th>merged</th>\n",
       "      <th>cdr3.len</th>\n",
       "      <th>antigen.epitope.len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1567</th>\n",
       "      <td>358</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSEGWHSYEQYF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ2-7*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSEGWHSYEQYF-KLGGALQAK</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1569</th>\n",
       "      <td>359</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASGLNIDGDEQFF</td>\n",
       "      <td>TRBV12-5*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV12-5*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASGLNIDGDEQFF-KLGGALQAK</td>\n",
       "      <td>14</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1575</th>\n",
       "      <td>362</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSVMLDSPLHF</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>TRBJ1-6*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV9*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSVMLDSPLHF-KLGGALQAK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>363</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSLESGFLSGYTF</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>TRBJ1-2*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*24:02</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>CYTWNQMNL</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV7-9*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLESGFLSGYTF-CYTWNQMNL</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1579</th>\n",
       "      <td>364</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSFTYRDTQYF</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>TRBJ2-3*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFTYRDTQYF-KLGGALQAK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42267</th>\n",
       "      <td>20708</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSLGGGRPQHF</td>\n",
       "      <td>TRBV7-8*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV7-8*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLGGGRPQHF-KLGGALQAK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42271</th>\n",
       "      <td>20710</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSHTSADEQFF</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>TRBJ2-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>RLRAEAQVK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSHTSADEQFF-RLRAEAQVK</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42275</th>\n",
       "      <td>20712</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSWRTRDNQPQHF</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>TRBJ1-5*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV28*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSWRTRDNQPQHF-KLGGALQAK</td>\n",
       "      <td>15</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42279</th>\n",
       "      <td>20714</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASSQDVRGSNSPLHF</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>TRBJ1-6*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV4-3*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSQDVRGSNSPLHF-KLGGALQAK</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42281</th>\n",
       "      <td>20715</td>\n",
       "      <td>TRB</td>\n",
       "      <td>CASRVGPEAFF</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>TRBJ1-1*01</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>HLA-A*03:01</td>\n",
       "      <td>B2M</td>\n",
       "      <td>MHCI</td>\n",
       "      <td>KLGGALQAK</td>\n",
       "      <td>...</td>\n",
       "      <td>TRBV6-1*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASRVGPEAFF-KLGGALQAK</td>\n",
       "      <td>11</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13271 rows × 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       complex.id gene              cdr3       v.segm      j.segm  \\\n",
       "1567          358  TRB    CASSEGWHSYEQYF   TRBV6-1*01  TRBJ2-7*01   \n",
       "1569          359  TRB    CASGLNIDGDEQFF  TRBV12-5*01  TRBJ2-1*01   \n",
       "1575          362  TRB     CASSVMLDSPLHF     TRBV9*01  TRBJ1-6*01   \n",
       "1577          363  TRB   CASSLESGFLSGYTF   TRBV7-9*01  TRBJ1-2*01   \n",
       "1579          364  TRB     CASSFTYRDTQYF    TRBV28*01  TRBJ2-3*01   \n",
       "...           ...  ...               ...          ...         ...   \n",
       "42267       20708  TRB     CASSLGGGRPQHF   TRBV7-8*01  TRBJ1-5*01   \n",
       "42271       20710  TRB     CASSHTSADEQFF   TRBV4-3*01  TRBJ2-1*01   \n",
       "42275       20712  TRB   CASSWRTRDNQPQHF    TRBV28*01  TRBJ1-5*01   \n",
       "42279       20714  TRB  CASSQDVRGSNSPLHF   TRBV4-3*01  TRBJ1-6*01   \n",
       "42281       20715  TRB       CASRVGPEAFF   TRBV6-1*01  TRBJ1-1*01   \n",
       "\n",
       "           species        mhc.a mhc.b mhc.class antigen.epitope  ...  \\\n",
       "1567   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "1569   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "1575   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "1577   HomoSapiens  HLA-A*24:02   B2M      MHCI       CYTWNQMNL  ...   \n",
       "1579   HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "...            ...          ...   ...       ...             ...  ...   \n",
       "42267  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "42271  HomoSapiens  HLA-A*03:01   B2M      MHCI       RLRAEAQVK  ...   \n",
       "42275  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "42279  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "42281  HomoSapiens  HLA-A*03:01   B2M      MHCI       KLGGALQAK  ...   \n",
       "\n",
       "       cdr3fix.vId cdr3fix.oldVEnd cdr3fix.oldVFixType  cdr3fix.oldVId  \\\n",
       "1567    TRBV6-1*01             NaN                 NaN             NaN   \n",
       "1569   TRBV12-5*01             NaN                 NaN             NaN   \n",
       "1575      TRBV9*01             NaN                 NaN             NaN   \n",
       "1577    TRBV7-9*01             NaN                 NaN             NaN   \n",
       "1579     TRBV28*01             NaN                 NaN             NaN   \n",
       "...            ...             ...                 ...             ...   \n",
       "42267   TRBV7-8*01             NaN                 NaN             NaN   \n",
       "42271   TRBV4-3*01             NaN                 NaN             NaN   \n",
       "42275    TRBV28*01             NaN                 NaN             NaN   \n",
       "42279   TRBV4-3*01             NaN                 NaN             NaN   \n",
       "42281   TRBV6-1*01             NaN                 NaN             NaN   \n",
       "\n",
       "      cdr3fix.oldJFixType cdr3fix.oldJId cdr3fix.oldJStart  \\\n",
       "1567                  NaN            NaN               NaN   \n",
       "1569                  NaN            NaN               NaN   \n",
       "1575                  NaN            NaN               NaN   \n",
       "1577                  NaN            NaN               NaN   \n",
       "1579                  NaN            NaN               NaN   \n",
       "...                   ...            ...               ...   \n",
       "42267                 NaN            NaN               NaN   \n",
       "42271                 NaN            NaN               NaN   \n",
       "42275                 NaN            NaN               NaN   \n",
       "42279                 NaN            NaN               NaN   \n",
       "42281                 NaN            NaN               NaN   \n",
       "\n",
       "                           merged cdr3.len antigen.epitope.len  \n",
       "1567     CASSEGWHSYEQYF-KLGGALQAK       14                   9  \n",
       "1569     CASGLNIDGDEQFF-KLGGALQAK       14                   9  \n",
       "1575      CASSVMLDSPLHF-KLGGALQAK       13                   9  \n",
       "1577    CASSLESGFLSGYTF-CYTWNQMNL       15                   9  \n",
       "1579      CASSFTYRDTQYF-KLGGALQAK       13                   9  \n",
       "...                           ...      ...                 ...  \n",
       "42267     CASSLGGGRPQHF-KLGGALQAK       13                   9  \n",
       "42271     CASSHTSADEQFF-RLRAEAQVK       13                   9  \n",
       "42275   CASSWRTRDNQPQHF-KLGGALQAK       15                   9  \n",
       "42279  CASSQDVRGSNSPLHF-KLGGALQAK       16                   9  \n",
       "42281       CASRVGPEAFF-KLGGALQAK       11                   9  \n",
       "\n",
       "[13271 rows x 57 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_human_10x[df_human_10x[\"antigen.epitope\"].isin(unique_10x)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_human_10x[df_human_10x[\"antigen.epitope\"].isin(unique_10x)].to_csv(PROJECT_ROOT / \"data/interim/vdjdb-2019-08-08/vdjdb-human-trb-mhci-10xunique-size.csv\", index=False, sep=\";\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "KLGGALQAK     12548\n",
       "RLRAEAQVK       407\n",
       "RTLNAWVKV        46\n",
       "AYAQKIFKI        39\n",
       "SLFNTVATLY       38\n",
       "FLASKIGRLV       34\n",
       "FLYALALLL        28\n",
       "LLDFVRFMGV       21\n",
       "KTWGQYWQV        16\n",
       "CYTWNQMNL        15\n",
       "MLDLQPETT        14\n",
       "YLLEMLWRL        13\n",
       "IMDQVPFSV        12\n",
       "KVLEYVIKV         8\n",
       "YLNDHLEPWI        7\n",
       "CLLWSFQTSA        7\n",
       "QPRAPIRPI         5\n",
       "SLFNTVATL         5\n",
       "CLLGTYTQDV        3\n",
       "SLYNTVATLY        2\n",
       "KVAELVHFL         1\n",
       "KLQCVDLHV         1\n",
       "RIAAWMATY         1\n",
       "Name: antigen.epitope, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_human_10x.loc[df_human_10x[\"antigen.epitope\"].isin(unique_10x), \"antigen.epitope\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_indices = df_human_10x[df_human_10x[\"antigen.epitope\"] == \"KLGGALQAK\"].sample(n=400, random_state=42).index\n",
    "drop_indices = df_human_10x[df_human_10x[\"antigen.epitope\"] == \"KLGGALQAK\"].index.difference(sample_indices)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_human_10x[df_human_10x[\"antigen.epitope\"].isin(unique_10x)].drop(drop_indices).to_csv(PROJECT_ROOT / \"data/interim/vdjdb-2019-08-08/vdjdb-human-trb-mhci-10xuniquedown-size.csv\", index=False, sep=\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.9 64-bit ('deepTCR': conda)",
   "language": "python",
   "name": "python36964bitdeeptcrconda38e3842162f941b1a4fa8a33c2d0e579"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
